BCK-7387  Update firestarter with diversion handling

resuna opened the pull request:
Update firestarter to detect diversions and set a diverted flag if they're found.

resuna committed:
Added check for diversions. TODO: loading destinations at startup, periodically cleaning destination cache, actually logging diversions

resuna committed:
TYpo- missing :

resuna committed:
make dest_history a global

resuna committed:
Can't pop value that's not in a dict.

resuna committed:
Perform diversion check on flightplan and flifo messages.

resuna committed:
Add "diversion" proc and update db refs to use "id" rather than "ident".

resuna committed:
aTypo

resuna committed:
Add diversion column = (None) to expected results

resuna committed:
Attempt to add a diversion to test suite.

resuna committed:
Just make "diverted" a flag like in trackstream.

resuna committed:
Fix to test, looking for final state

resuna committed:
Force diversion for every row

resuna committed:
Wipe debug code.

resuna committed:
Re-order functions for clarity.

resuna committed:
Look for new flights that haven't had ETMS messages yet.

resuna committed:
Change diverted to sa.Boolean

NasaGeek commented on db-updater/main.py:
```diff
@@ -480,6 +491,54 @@ def process_keepalive_message(data: dict) -> None:
+    return
+
+
+def check_for_diversions(data: dict) -> None:
+    """ETMS message, check destination"""
+    if "dest" in data and "id" in data:
+        global dest_history
+
+        dest = data.get("dest")
+        if dest == "":
+            return
+
+        flight = data.get("id")
+        if flight == "":
+            return
+
+        if flight in dest_history:
+            orig_dest = dest_history.get(flight)
+            if orig_dest != "":
+                if orig_dest != dest:
+                    data["diverted"] = True
+
+        dest_history[flight] = dest
+
+    return
```
```suggestion
def check_for_diversions(data: dict) -> None:
    """ETMS message, check destination"""
    global dest_history

    dest = data.get("dest")
    flight = data.get("id")
    orig_dest = dest_history.get(flight)

    if dest and flight:
        if orig_dest and orig_dest != dest:
            data["diverted"] = True

        dest_history[flight] = dest

    return
```

Because `dict.get()` won't fail and Python makes it easy to broadly check if a variable is "truthy", we can simplify this function.

NasaGeek commented on db-updater/main.py:
```diff
@@ -480,6 +491,54 @@ def process_keepalive_message(data: dict) -> None:
+
+def check_for_new_destination(data: dict) -> None:
+    """Look for a new destination that hasn't been seen by an ETMS message yet"""
+    if "dest" in data and "id" in data:
+        global dest_history
+
+        flight = data.get("id")
+        if flight == "" or flight in dest_history:
+            return
+
+        dest = data.get("dest")
+        if dest == "":
+            return
+
+        dest_history[flight] = dest
+
+    return
```
```suggestion
    global dest_history
    dest = data.get("dest")
    flight = data.get("id")

    if dest and flight and flight not in dest_history:
        dest_history[flight] = dest

    return
```
Similar simplification as above.

NasaGeek reviewed (commented):
Some minor nits on use of `dict.get()` and leveraging truthiness.

resuna committed:
We know id is present, so don't need .get()

Co-authored-by: Chris Roberts <NasaGeek@users.noreply.github.com>

resuna committed:
We know id is present, so don't need .get()

Co-authored-by: Chris Roberts <NasaGeek@users.noreply.github.com>

resuna committed:
Use python idioms.

Because dict.get() won't fail and Python makes it easy to broadly check if a variable is "truthy", we can simplify this function.

Co-authored-by: Chris Roberts <NasaGeek@users.noreply.github.com>

resuna committed:
Use python idioms.

Because dict.get() won't fail and Python makes it easy to broadly check if a variable is "truthy", we can simplify this function.

Co-authored-by: Chris Roberts <NasaGeek@users.noreply.github.com>
